File: kicad_fdtd/voxel.py
"""Assign every Yee E-edge to a conductor, exactly as the solver will use it,
and check for merged conductors BEFORE solving.

Yee layout on node lines x[i], y[j], z[k]:
  Ex(i, j, k) is the edge from node (i, j, k) to (i+1, j, k), midpoint (mx[i], y[j], z[k])
  Ey(i, j, k)                        (i, j+1, k)            (x[i], my[j], z[k])
  Ez(i, j, k)                        (i, j, k+1)            (x[i], y[j], mz[k])
Copper sheets sit on z planes: an Ex/Ey edge on that plane is metal if its
midpoint is inside a polygon of that layer. Via barrels are solid cylinders:
an edge is metal if its midpoint is within r of the axis and inside the
barrel's z range. The result is one integer net id per edge (-1 = none).
"""
import numpy as np
import shapely
from shapely.geometry import box


def assign(bd, lines, tie_nets=None, skip_via_near=(), window=None):
    tie = tie_nets or {}
    x, y, z = lines["x"], lines["y"], lines["z"]
    nx, ny, nz = len(x), len(y), len(z)
    mx, my, mz = (x[:-1] + x[1:]) / 2, (y[:-1] + y[1:]) / 2, (z[:-1] + z[1:]) / 2
    ex = np.full((nx - 1, ny, nz), -1, np.int16)
    ey = np.full((nx, ny - 1, nz), -1, np.int16)
    ez = np.full((nx, ny, nz - 1), -1, np.int16)
    names = []

    def nid(net):
        net = tie.get(net, net)
        if net not in names:
            names.append(net)
        return names.index(net)
    zidx = {layer: int(np.argmin(abs(z - bd.z[layer]))) for layer in bd.layers}
    for layer in bd.layers:
        assert abs(z[zidx[layer]] - bd.z[layer]) < 1e-6, "copper plane %s is not on a z line" % layer
    win = box(*window) if window else None
    XX, YY = np.meshgrid(mx, y, indexing="ij")
    XY, YX = np.meshgrid(x, my, indexing="ij")
    for layer in bd.layers:
        k = zidx[layer]
        for net, kind, P in bd.polys(layer):
            if win is not None and not P.intersects(win):
                continue
            b = P.bounds
            ix = np.where((mx >= b[0] - 1e-9) & (mx <= b[2] + 1e-9))[0]
            jy = np.where((y >= b[1] - 1e-9) & (y <= b[3] + 1e-9))[0]
            if len(ix) and len(jy):
                sub = shapely.contains_xy(P, XX[np.ix_(ix, jy)].ravel(), YY[np.ix_(ix, jy)].ravel()).reshape(len(ix), len(jy))
                if sub.any():
                    blk = ex[np.ix_(ix, jy, [k])][:, :, 0]
                    blk[sub] = nid(net)
                    ex[np.ix_(ix, jy, [k])] = blk[:, :, None]
            ix = np.where((x >= b[0] - 1e-9) & (x <= b[2] + 1e-9))[0]
            jy = np.where((my >= b[1] - 1e-9) & (my <= b[3] + 1e-9))[0]
            if len(ix) and len(jy):
                sub = shapely.contains_xy(P, XY[np.ix_(ix, jy)].ravel(), YX[np.ix_(ix, jy)].ravel()).reshape(len(ix), len(jy))
                if sub.any():
                    blk = ey[np.ix_(ix, jy, [k])][:, :, 0]
                    blk[sub] = nid(net)
                    ey[np.ix_(ix, jy, [k])] = blk[:, :, None]
    n_via = 0
    for v in bd.vias():
        if win is not None and not win.contains(shapely.geometry.Point(v["x"], v["y"])):
            continue
        if any(np.hypot(v["x"] - px, v["y"] - py) < 0.6 for px, py in skip_via_near):
            continue
        k = nid(v["net"])
        r = v["r"]
        zb, zt = min(v["zb"], v["zt"]), max(v["zb"], v["zt"])
        kz = np.where((z >= zb - 1e-9) & (z <= zt + 1e-9))[0]            # planes the barrel crosses
        kzm = np.where((mz >= zb - 1e-9) & (mz <= zt + 1e-9))[0]         # z-edges inside the barrel
        ix = np.where((mx >= v["x"] - r) & (mx <= v["x"] + r))[0]; jy = np.where((y >= v["y"] - r) & (y <= v["y"] + r))[0]
        if len(ix) and len(jy):
            d = np.hypot(XX[np.ix_(ix, jy)] - v["x"], YY[np.ix_(ix, jy)] - v["y"]) <= r
            for kk in kz:
                blk = ex[np.ix_(ix, jy, [kk])][:, :, 0]; blk[d] = k; ex[np.ix_(ix, jy, [kk])] = blk[:, :, None]
        ix = np.where((x >= v["x"] - r) & (x <= v["x"] + r))[0]; jy = np.where((my >= v["y"] - r) & (my <= v["y"] + r))[0]
        if len(ix) and len(jy):
            d = np.hypot(XY[np.ix_(ix, jy)] - v["x"], YX[np.ix_(ix, jy)] - v["y"]) <= r
            for kk in kz:
                blk = ey[np.ix_(ix, jy, [kk])][:, :, 0]; blk[d] = k; ey[np.ix_(ix, jy, [kk])] = blk[:, :, None]
        ix = np.where((x >= v["x"] - r) & (x <= v["x"] + r))[0]; jy = np.where((y >= v["y"] - r) & (y <= v["y"] + r))[0]
        if len(ix) and len(jy) and len(kzm):
            NX, NY = np.meshgrid(x[ix], y[jy], indexing="ij")
            d = np.hypot(NX - v["x"], NY - v["y"]) <= r
            for kk in kzm:
                blk = ez[np.ix_(ix, jy, [kk])][:, :, 0]; blk[d] = k; ez[np.ix_(ix, jy, [kk])] = blk[:, :, None]
        n_via += 1
    return {"ex": ex, "ey": ey, "ez": ez, "names": names, "n_via": n_via}


def merge_check(vox, lines, nets_of_interest, tie_nets=None):
    """Nodes touched by edges of two different conductors = a short in the
    solver. Returns [(netA, netB, x, y, z)] for nodes involving nets_of_interest."""
    tie = tie_nets or {}
    ex, ey, ez, names = vox["ex"], vox["ey"], vox["ez"], vox["names"]
    x, y, z = lines["x"], lines["y"], lines["z"]
    nx, ny, nz = len(x), len(y), len(z)
    NONE = -1
    # per node: collect ids of the up to 6 touching edges
    ids = [np.full((nx, ny, nz), NONE, np.int16) for _ in range(6)]
    ids[0][1:, :, :] = ex; ids[1][:-1, :, :] = ex
    ids[2][:, 1:, :] = ey; ids[3][:, :-1, :] = ey
    ids[4][:, :, 1:] = ez; ids[5][:, :, :-1] = ez
    stack = np.stack(ids, axis=-1)
    mn = np.where(stack >= 0, stack, 32000).min(axis=-1)
    mx_ = stack.max(axis=-1)
    bad = (mx_ >= 0) & (mn < 32000) & (mn != mx_)
    out = []
    watch = {tie.get(n, n) for n in nets_of_interest}
    for i, j, k in zip(*np.where(bad)):
        s = sorted({int(v) for v in stack[i, j, k] if v >= 0})
        nm = [names[v] for v in s]
        if any(n in watch for n in nm):
            out.append((nm[0], nm[1], float(x[i]), float(y[j]), float(z[k])))
    return out

File: docs/app/voxel.js
// Exact Yee-edge assignment (port of kicad_fdtd/voxel.py): every E edge whose midpoint lies
// inside a copper polygon on that layer's z plane belongs to that conductor; via barrels
// assign the edges inside their cylinder on every plane and the z edges they cross.
// merge_check: nodes touched by edges of two different conductors = shorts in the solver.
import { scanPoly, lowerBound, upperBound } from "./geom.js";

export function assign(bd, lines, tie = {}, skipViaNear = [], window = null) {
  const x = lines.x, y = lines.y, z = lines.z;
  const nx = x.length, ny = y.length, nz = z.length;
  const mx = new Float64Array(nx - 1), my = new Float64Array(ny - 1), mz = new Float64Array(nz - 1);
  for (let i = 0; i < nx - 1; i++) mx[i] = (x[i] + x[i + 1]) / 2;
  for (let j = 0; j < ny - 1; j++) my[j] = (y[j] + y[j + 1]) / 2;
  for (let k = 0; k < nz - 1; k++) mz[k] = (z[k] + z[k + 1]) / 2;
  const ex = new Int16Array((nx - 1) * ny * nz).fill(-1);      // index (i*ny + j)*nz + k
  const ey = new Int16Array(nx * (ny - 1) * nz).fill(-1);      // (i*(ny-1) + j)*nz + k
  const ez = new Int16Array(nx * ny * (nz - 1)).fill(-1);      // (i*ny + j)*(nz-1) + k
  const names = [];
  const nid = net => { net = tie[net] || net; let i = names.indexOf(net); if (i < 0) { names.push(net); i = names.length - 1; } return i; };
  const zidx = {};
  for (const layer of bd.layers) {
    let best = 0; for (let k = 1; k < nz; k++) if (Math.abs(z[k] - bd.z[layer]) < Math.abs(z[best] - bd.z[layer])) best = k;
    if (Math.abs(z[best] - bd.z[layer]) > 1e-6) throw new Error(`copper plane ${layer} is not on a z line`);
    zidx[layer] = best;
  }
  const inWin = P => !window || !(P.bbox[2] < window[0] || P.bbox[0] > window[2] || P.bbox[3] < window[1] || P.bbox[1] > window[3]);
  for (const layer of bd.layers) {
    const k = zidx[layer];
    for (const P of bd.polys(layer)) {
      if (!inWin(P)) continue;
      const id = nid(P.net);
      scanPoly(P, mx, y, (i, j) => { ex[(i * ny + j) * nz + k] = id; });
      scanPoly(P, x, my, (i, j) => { ey[(i * (ny - 1) + j) * nz + k] = id; });
    }
  }
  let nVia = 0;
  for (const v of bd.vias()) {
    if (window && !(v.x >= window[0] && v.x <= window[2] && v.y >= window[1] && v.y <= window[3])) continue;
    if (skipViaNear.some(([px, py]) => Math.hypot(v.x - px, v.y - py) < 0.6)) continue;
    const id = nid(v.net), r = v.r;
    const zb = Math.min(v.zb, v.zt), zt = Math.max(v.zb, v.zt);
    const kz = [], kzm = [];
    for (let k = 0; k < nz; k++) if (z[k] >= zb - 1e-9 && z[k] <= zt + 1e-9) kz.push(k);
    for (let k = 0; k < nz - 1; k++) if (mz[k] >= zb - 1e-9 && mz[k] <= zt + 1e-9) kzm.push(k);
    // ex edges: midpoints (mx[i], y[j]) inside the circle
    for (let i = lowerBound(mx, v.x - r); i < upperBound(mx, v.x + r); i++)
      for (let j = lowerBound(y, v.y - r); j < upperBound(y, v.y + r); j++)
        if (Math.hypot(mx[i] - v.x, y[j] - v.y) <= r) for (const k of kz) ex[(i * ny + j) * nz + k] = id;
    for (let i = lowerBound(x, v.x - r); i < upperBound(x, v.x + r); i++)
      for (let j = lowerBound(my, v.y - r); j < upperBound(my, v.y + r); j++)
        if (Math.hypot(x[i] - v.x, my[j] - v.y) <= r) for (const k of kz) ey[(i * (ny - 1) + j) * nz + k] = id;
    for (let i = lowerBound(x, v.x - r); i < upperBound(x, v.x + r); i++)
      for (let j = lowerBound(y, v.y - r); j < upperBound(y, v.y + r); j++)
        if (Math.hypot(x[i] - v.x, y[j] - v.y) <= r) for (const k of kzm) ez[(i * ny + j) * (nz - 1) + k] = id;
    nVia++;
  }
  return { ex, ey, ez, names, nVia, nx, ny, nz };
}

export function mergeCheck(vox, lines, netsOfInterest, tie = {}) {
  const { ex, ey, ez, names, nx, ny, nz } = vox;
  const x = lines.x, y = lines.y, z = lines.z;
  const watch = new Set(netsOfInterest.map(n => tie[n] || n));
  const out = [];
  const seen = new Set();
  const consider = (i, j, k, ids) => {
    let mn = 32000, mx_ = -1;
    for (const v of ids) if (v >= 0) { if (v < mn) mn = v; if (v > mx_) mx_ = v; }
    if (mx_ < 0 || mn === mx_) return;
    const s = [...new Set(ids.filter(v => v >= 0))].sort((a, b) => a - b).map(v => names[v]);
    if (s.some(n => watch.has(n))) { const key = i + "," + j + "," + k; if (!seen.has(key)) { seen.add(key); out.push([s[0], s[1], x[i], y[j], z[k]]); } }
  };
  const gx = (i, j, k) => (i < 0 || i >= nx - 1) ? -1 : ex[(i * ny + j) * nz + k];
  const gy = (i, j, k) => (j < 0 || j >= ny - 1) ? -1 : ey[(i * (ny - 1) + j) * nz + k];
  const gz = (i, j, k) => (k < 0 || k >= nz - 1) ? -1 : ez[(i * ny + j) * (nz - 1) + k];
  for (let i = 0; i < nx; i++) for (let j = 0; j < ny; j++) for (let k = 0; k < nz; k++) {
    const a = gx(i - 1, j, k), b = gx(i, j, k), c = gy(i, j - 1, k), d = gy(i, j, k), e = gz(i, j, k - 1), f = gz(i, j, k);
    if (a < 0 && b < 0 && c < 0 && d < 0 && e < 0 && f < 0) continue;
    consider(i, j, k, [a, b, c, d, e, f]);
  }
  return out;
}
